[enh] Helpers for common IO operations

alexAubin opened the pull request:
### Status

- Kinda stable now (with tests). Need feedback !
- Wondering if we should put this in the Moulinette

### Problem

Follow-up of a discussion we had two or three times already : we keep needing to read / write json files, change file permissions or downloading stuff at various places in the code. This lead to either duplicated code, or different pieces of code that don't handle exceptions consistently. The current situation also discourage to handle exceptions correctly because there are simply too many of them to check (e.g. when doing a `requests.get()`)

### Proposed solution

Introduce a `io.py` that contains low-level functions for all common operations such as : 
- read/write files
- read/write json
- download text data
- change file permissions
- execute ~raw bash commands
- ???

Psycojoker commented:
Haven't dig into it too much but looks like a really good start :+1:

opi commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+    Keyword argument:
+        file_path -- Path to the json file
+    """
+    assert isinstance(file_path, str)
```
assertion is already done in _read_file()

opi commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+
+    if os.path.exists(file_path):
+        try:
+            os.remove(file_path)
```
`os.remove` will probably raise an IOError or OSError exception that could be catched.

opi commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+
+    # Download file
+    try:
+        r = requests.get(url, timeout=30)
```
Timeout could be passed as an optional argument to  _download_text_data().

Psycojoker replied:
According to requests documentation we should **always** have a timeout, so only the value should be modifiable http://docs.python-requests.org/en/latest/user/quickstart/#timeouts

opi commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+                              m18n.n('io_download_bad_status_code',
+                                     url=url, code=r.status_code))
+
+    return r.text
```
Maybe we could provide a wrapper around `_download_text_data()` that download, parse and return valid JSON ?

julienmalik commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+        raise MoulinetteError(errno.EROENT,
+                              m18n.n('io_no_such_file', file=file_path))
+
+    uid = pwd.getpwnam(user).pw_uid
```
what's happening when the user or group does not exists ?

julienmalik commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+    assert isinstance(permissions, int)
+
+    # Check file exist
+    if not os.path.isfile(file_path):
```
imho this function should support directories too, transparently

julienmalik commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+    You should have received a copy of the GNU Affero General Public License
+    along with this program; if not, see http://www.gnu.org/licenses
+
+    yunohost_certificate.py
```
copypasta

julienmalik commented on src/yunohost/io.py:
```diff
@@ -0,0 +1,222 @@
+
+    yunohost_certificate.py
+
+    Manage certificates, in particular Let's encrypt
```
copypasta

zamentur commented:
May be this code should be move into moulinette repository ?

alexAubin commented:
>May be this code should be move into moulinette repository ?

This is an interesting idea considering that this is pretty independent from yunohost

alexAubin commented:
I added tests for all cases I can think about and fixed / stabilized the functions accordingly. Code is ready for reviews. Feel free to suggest additional functions/helpers or different API

alexAubin commented:
Ok I just realized there's a "moulinette/utils/filesystem.py" which implements part of this stuff..

alexAubin commented:
Closing because moved to https://github.com/YunoHost/moulinette/pull/141
